flink-runtime/src/main/java/org/apache/flink/runtime/state/CachedCheckpointStreamFactory.java

```diff
@@ -0,0 +1,195 @@
+			cachedOut = cache.createOutputStream(checkpointID, handleID, placeholder);
+		}
+		CheckpointStateOutputStream remoteOut = null;
+		if (!placeholder) {
+			remoteOut = remoteFactory.createCheckpointStateOutputStream(checkpointID, timestamp);
+		}
+		CachedCheckpointStateOutputStream output = new CachedCheckpointStateOutputStream(cachedOut, remoteOut);
+		return output;
+	}
+
+	@Override
+	public CheckpointStateOutputStream createCheckpointStateOutputStream(long checkpointID, long timestamp) throws Exception {
+		LOG.warn("create output stream which is not cacheable.");
+		return remoteFactory.createCheckpointStateOutputStream(checkpointID, timestamp);
+	}
+
+	@Override
+	public void close() throws Exception {
+		remoteFactory.close();
+	}
+
+	/**
+	 * A hybrid checkpoint output stream which write data to both remote end and local end,
+	 * writing data locally failed won't stop writing to remote. This hybrid output stream
+	 * will return a {@link CachedStreamStateHandle} in closeAndGetHandle(), it can be used for read data locally.
+	 */
+	public static class CachedCheckpointStateOutputStream extends CheckpointStateOutputStream {
+
+		private CachedOutputStream cacheOut = null;
+		private CheckpointStateOutputStream remoteOut = null;
+
+		public CachedCheckpointStateOutputStream(CachedOutputStream cacheOut, CheckpointStateOutputStream remoteOut) {
+			this.cacheOut = cacheOut;
+			this.remoteOut = remoteOut;
+		}
+
+		@Override
+		public StreamStateHandle closeAndGetHandle() throws IOException {
+			if (cacheOut != null) {
+				// finalize cache data
+				StateHandleID cacheId = cacheOut.getCacheID();
+				cacheOut.end();
+
+				StreamStateHandle remoteHandle;
+				if (remoteOut != null) {
+					remoteHandle = remoteOut.closeAndGetHandle();
+				} else {
+					remoteHandle = new PlaceholderStreamStateHandle(cacheId);
+				}
+				return new CachedStreamStateHandle(cacheId, remoteHandle);
+			} else {
+				if (remoteOut != null) {
+					return remoteOut.closeAndGetHandle();
+				} else {
+					return null;
+				}
+			}
+		}
+
+		@Override
+		public long getPos() throws IOException {
+			return remoteOut != null ? remoteOut.getPos() :-1L;
+		}
+
+		@Override
+		public void write(int b) throws IOException {
+			// write to local
+			if (cacheOut != null) {
+				try {
+					cacheOut.write(b);
+				} catch (Exception e) {
+					//discard
+					cacheOut.discard();
+				}
+			}
+
+			// write to remote
+			if (remoteOut != null) {
+				remoteOut.write(b);
+			}
+		}
+
+		@Override
+		public void write(byte[] b, int off, int len) throws IOException {
+			// write to local
+			if (cacheOut != null) {
+				try {
+					cacheOut.write(b, off, len);
+				} catch (Exception e) {
+					//discard
+					cacheOut.discard();
+				}
+			}
+
+			// write to remote
+			if (remoteOut != null) {
+				remoteOut.write(b, off, len);
+			}
+		}
+
+		@Override
+		public void flush() throws IOException {
+			if (cacheOut != null) {
+				cacheOut.flush();
```

Comment:
I found this a bit inconsistent: why not also catching exceptions from `cacheOut` here and discard the stream? In general, I would suggest to write this class in a way that failures on `cacheOut` to not result in a checkpoint failure (or you can make this configurable). As long as writing to the primary (remote) stream works, I would consider the checkpoint as ok, caching is more of a bonus.

Reply:
You're right, What My intend is what you described but I missed it here, I should handle it like above code, and wrapped up the catchOut operation with try catch block.